pkg/controller/deployment/deployment_controller.go

```diff
@@ -630,9 +630,73 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 		// Cant scale down further
 		return false, nil
 	}
+
+	// Check the expectations of deployment before reconciling
+	dKey, err := controller.KeyFunc(&deployment)
+	if err != nil {
+		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	}
+	if expectationsCheck && !dc.podExpectations.SatisfiedExpectations(dKey) {
+		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+		return false, nil
+	}
+
+	// Clean up all unhealthy replicas, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment)
+	if err != nil {
+		return false, nil
+	}
+
+	// Scale down old rcs, need check maxUnavailable to ensure we can scale down
+	scaledDown, err := dc.scaleDownOldRCsForRollingUpdate(allRCs, oldRCs, deployment)
+	if err != nil {
+		return false, nil
+	}
+
+	totalScaledDown := cleanupCount + scaledDown
+	if expectationsCheck {
+		dc.podExpectations.ExpectDeletions(dKey, totalScaledDown)
+	}
+
+	return totalScaledDown > 0, nil
+}
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment) (int, error) {
```

Comment:
Thanks for taking a stab at this. 

The main disadvantage of this approach is that it will resize RCs to eliminate all pods that happen to be unhealthy at the moment, causing the rollout to advance unpredictably. This was the original problem we had that motivated us to use multiple RCs to perform rolling update. If we wanted this behavior, it would be simpler to just use a single RC, update its template in place, and delete the old/unhealthy pods, allowing the RC to replace them using the new template. Background is in #1353.

So, we need to limit the number of old pods we delete by resizing old RCs.

I'll describe this in terms of absolute counts for maxUnavailable and maxSurge.

Let's say the user rolled out a new revision, which deleted maxUnavailable instances of the old revision and added maxUnavailable + maxSurge instances of the new. 

However, the new revision crash loops and those pods never become ready. The user then decides to rollback.

Now the user has replicas-maxUnavailable (available) instances of the "new new" (rolled back) revision and maxUnavailable+maxSurge (unavailable) instances of the "old new" revision. We want this rollback to make progress. However, coincidentally, a node fails and causes one of the "new new" pods to fail and be recreated, but it is Pending while its image is pulled to a new node, so there are replicas-maxUnavailable-1 available instances of the "new new" revision, causing the Deployment to be over its maxUnavailable threshold. 

Intuitively, we should be able to revert maxUnavailable of the "old new" pods and delete maxSurge "old new" pods as the reverted pods become available. 

Invariants:
* There should never be more than replicas+maxSurge pods total.
* There should never be more than replicas pods of the "new new" RC.
* There shouldn't be more than maxUnavailable+maxSurge pods of the "new new" RC unavailable (though it could happen due to circumstances outside the control of the Deployment controller).
* There shouldn't be more than replicas pods available total (though there may be temporarily).

If there are fewer than maxUnavailable+maxSurge pods of the "new new" RC unavailable after having created at least maxSurge pods, then that is the budget that can be used to prune unavailable instances from old RCs.

This is not the best example because it doesn't leave pods that we don't want to delete, but hopefully this makes sense.